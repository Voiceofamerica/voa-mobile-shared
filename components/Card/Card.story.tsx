
import 'react-native'
import * as React from 'react'
import * as RX from 'reactxp'
import { storiesOf, linkTo } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'

import Card, { ArticleBlurb } from './Card'

const testArticle: ArticleBlurb = {
  image: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  minor: '15:08',
  title: 'Pressident Trump: From North Korea to Iran, we confront dangerous governments',
}

const testArticle2: ArticleBlurb = {
  image: 'https://gdb.voanews.com/01D5530E-3642-4CBF-957E-AE990ABE9167_cx0_cy14_cw0_w650_r1.jpg',
  minor: '21:55',
  title: 'At least seven people were killed in a police vehicle in Balochistan, Pakistan',
}

storiesOf('Card', module)
  .add('display', () => (
    <RX.View style={{ flexDirection: 'row' }}>
      <Card onPress={action('Card.onPress')} blurb={testArticle} />
    </RX.View>
  ))
  .add('display two', () => (
    <RX.View style={{ flexDirection: 'row' }}>
      <Card onPress={action('Card.onPress')} blurb={testArticle} factor={2} />
      <Card onPress={action('Card.onPress')} blurb={testArticle2} factor={2} />
    </RX.View>
  ))
