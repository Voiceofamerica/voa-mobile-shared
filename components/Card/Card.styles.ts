
import * as RN from 'react-native'
import * as RX from 'reactxp'

export default {
  container: RX.Styles.createViewStyle({
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    borderColor: 'rgba(128, 128, 128, 0.5)',
    borderWidth: 1,
    backgroundColor: '#DDD',
    margin: 5,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 10 },
    shadowOpacity: 10,
    shadowRadius: 10,
    borderRadius: 5,
    elevation: 6,
    overflow: 'hidden',
  }),
  image: RX.Styles.createImageStyle({
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  }),
  gradientContainer: RX.Styles.createViewStyle({
    flexDirection: 'row',
  }),
  gradient: RX.Styles.createViewStyle({
    flex: 1,
    padding: 8,
    paddingTop: 30,
  }) as RN.ViewStyle,
  minorText: RX.Styles.createTextStyle({
    color: '#FFF',
    paddingBottom: 5,
    backgroundColor: 'transparent',
  }),
  titleText: RX.Styles.createTextStyle({
    color: '#FFF',
    fontWeight: 'bold',
    backgroundColor: 'transparent',
  }),
}
