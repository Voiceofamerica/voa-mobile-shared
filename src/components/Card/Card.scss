
@import '../../commonStyles/index.scss';

.card {
  @include touchable();

  flex: 1;
  position: relative;
  display: flex;
  background-size: cover;
  background-position: center;
  flex-direction: column;
  justify-content: flex-end;
  overflow: hidden;
  user-select: none;
  border-bottom: 1vw solid $voa-red;

  .imageContainer {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }

  .imageStyle {
    display: block;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  .gradient {
    background: linear-gradient(transparent, rgba(0, 0, 0, 0.9));
    padding: 1.5vw;
    padding-top: 30%;
    z-index: 1;
  }

  .titleText {
    color: white;
    text-shadow: 0.5vh 0.5vw 0.5vw $accent-grey;
    font-weight: bold;
    width: 50%;
    padding: 0 0 2.5vh 5vw;
    font-size: 20px;
  }
}
