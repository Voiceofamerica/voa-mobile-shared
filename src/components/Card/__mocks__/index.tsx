
import * as React from 'react'

export default (props) => (
  React.createElement('card', { props }, props.children)
)

export const getHeight = () => {
  return 10
}

export const CARD_PADDING = 11
