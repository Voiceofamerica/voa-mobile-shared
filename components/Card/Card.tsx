
import * as React from 'react'
import * as moment from 'moment'
import { Subscription } from 'rxjs/Subscription'

import { resize } from '../../helpers/windowHelper'

import ResilientImage from '../ResilientImage'
import ArticleBlurb from '../../types/ArticleBlurb'

import { card, imageStyle, imageContainer, gradient, minorText, titleText } from './Card.scss'

export interface Props {
  onPress: () => void
  blurb: ArticleBlurb
  factor?: number
}

export interface State {
  windowWidth: number
}

const HEIGHT_RATIO = 3 / 4
const TITLE_RATIO = 1 / 20
const MINOR_RATIO = 1 / 30
const TEXT_POWER = 1 / 2

class Card extends React.Component<Props, State> {
  state = {
    windowWidth: window.innerWidth,
  }

  private _self: HTMLDivElement
  private _resizeSub: Subscription

  componentDidMount () {
    this._resizeSub = resize()
      .map(({ width }) => width)
      .subscribe(windowWidth => this.setState({ windowWidth }))
  }

  componentWillUnmount () {
    this._resizeSub.unsubscribe()
  }

  render () {
    const { windowWidth } = this.state
    const { onPress, blurb: { image, pubDate, title }, factor = 1 } = this.props
    const width = this._self ? this._self.clientWidth : windowWidth / factor

    const height = width * HEIGHT_RATIO

    const textRatio = Math.pow(windowWidth / (width + 10), TEXT_POWER)
    const titleSize = width * TITLE_RATIO * textRatio
    const minorSize = width * MINOR_RATIO * textRatio

    return (
      <div
        ref={el => this._self = el}
        className={card}
        style={{ height }}
        onClick={onPress}>
        <ResilientImage className={imageContainer} src={image && image.url} alwaysShow />
        <div className={gradient}>
          <div className={minorText} style={{ fontSize: minorSize }}>
            {moment(pubDate).fromNow()}
          </div>
          <div className={titleText} style={{ fontSize: titleSize }}>
            {title}
          </div>
        </div>
      </div>
    )
  }
}

export default Card
