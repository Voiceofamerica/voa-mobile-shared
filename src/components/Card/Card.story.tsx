
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import ArticleBlurb from '../../types/ArticleBlurb'

import Card from './Card'

const testArticle = {
  imageUrl: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const testArticle2 = {
  imageUrl: 'https://gdb.voanews.com/01D5530E-3642-4CBF-957E-AE990ABE9167_cx0_cy14_cw0_w650_r1.jpg',
  pubDate: '21:55',
  title: 'At least seven people were killed in a police vehicle in Balochistan, Pakistan',
}

storiesOf('Card', module)
  .add('display', () => (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Card onPress={action('Card.onPress')} { ...testArticle } />
    </div>
  ))
  .add('display two', () => (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Card onPress={action('Card.onPress')} { ...testArticle } factor={2} />
      <Card onPress={action('Card.onPress')} { ...testArticle2 } factor={2} />
    </div>
  ))
  .add('display with icon', () => (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Card
        onPress={action('Card.onPress')}
        { ...testArticle }
        title={(
          <span><img src={require('./headphones.svg')} />{testArticle.title}</span>
        )}
      />
    </div>
  ))
