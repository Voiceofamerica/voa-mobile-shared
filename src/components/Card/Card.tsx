
import * as React from 'react'

import { toRGBAstring } from '../../helpers/colorHelper'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'
import ResilientImage from '../ResilientImage'
import SvgIcon from '../SvgIcon'

import { card, imageContainer, gradient, titleText } from './Card.scss'

export interface Props extends ThemeProps {
  title: string
  onPress?: () => void
  imageUrl?: string | null
  icon?: JSX.Element | string
  className?: string
  style?: React.CSSProperties
}

const IMAGE_RATIO = 9 / 16

export const getHeight = (renderWidth = window.innerWidth, imageRatio = IMAGE_RATIO) => {
  return renderWidth * imageRatio
}

// 4px border-bottom + 7px margin-bottom
export const CARD_PADDING = 11

@themed
class Card extends React.Component<Props> {
  render () {
    const { onPress, icon, title, imageUrl, className, style, theme = DEFAULT_THEME } = this.props
    const {
      cardGradientColor,
      cardTitleColor,
      cardBorderColor,
    } = theme

    const gradientEnd = toRGBAstring(cardGradientColor, 0.9)
    const gradientBackground = `linear-gradient(transparent, ${gradientEnd})`

    const hydratedIcon = typeof icon === 'string'
                       ? <SvgIcon src={icon} />
                       : icon

    return (
      <div
        className={`${card} ${className}`}
        style={{
          ...style,
          borderBottomColor: cardBorderColor,
        }}
        onClick={onPress}>
        <ResilientImage className={imageContainer} src={imageUrl} alwaysShow showSpinner />
        <div className={gradient} style={{ background: gradientBackground }}>
          <div className={titleText} style={{ color: cardTitleColor }}>
            <span>{hydratedIcon} {title}</span>
          </div>
        </div>
      </div>
    )
  }
}

export default Card
