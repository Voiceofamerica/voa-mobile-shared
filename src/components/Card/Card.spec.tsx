
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import Card from './Card'

jest.mock('../../helpers/windowHelper')
jest.mock('../ResilientImage')

describe('<Card />', () => {
  describe('shapshots', () => {
    it('should render with the given title', () => {
      const element = create((
        <Card title='myTitle' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given image', () => {
      const element = create((
        <Card title='myTitle' imageUrl='myImage' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given minor text', () => {
      const element = create((
        <Card title='myTitle' minorText='This is some minor text' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the a height determined by its factor', () => {
      const element = create((
        <Card title='myTitle' factor={3} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given icon shown', () => {
      const element = create((
        <Card title='myTitle' icon={<i className='mdi mdi-monitor' />} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given className', () => {
      const element = create((
        <Card title='myTitle' className='myClass' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given style', () => {
      const element = create((
        <Card title='myTitle' style={{ top: 0, left: 0, right: 0, bottom: 0 }} />
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onPress', () => {
    it('should be called whenever the element is clicked', () => {
      const onPress = jest.fn()

      const item = shallow((
        <Card title='myTitle' onPress={onPress} />
      ))

      item.simulate('click')

      expect(onPress).toHaveBeenCalledTimes(1)
    })

    it('should not throw an error if not provided and the element is clicked', () => {
      const item = shallow((
        <Card title='myTitle' />
      ))

      item.simulate('click')
    })
  })

  describe('resize', () => {
    it('should change height when window is resized', () => {
      const windowHelper = require('../../helpers/windowHelper')

      const item = shallow((
        <Card title='myTitle' />
      ))

      const originalWidth = item.state().windowWidth
      const expected = originalWidth + 400
      windowHelper.setWindowWidth(expected)
      const actual = item.state().windowWidth

      expect(actual).not.toEqual(originalWidth)
      expect(expected).toEqual(actual)
    })

    it('should null its subscription upon unmount', () => {
      const windowHelper = require('../../helpers/windowHelper')

      const item = shallow((
        <Card title='myTitle' />
      ))

      const card = item.instance() as Card
      expect(card._resizeSub).not.toBeNull()

      item.unmount()

      expect(card._resizeSub).toBeNull()
    })
  })
})
