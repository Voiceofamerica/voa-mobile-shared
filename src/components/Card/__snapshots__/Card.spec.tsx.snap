// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> shapshots should render with the given className 1`] = `
<div
  className="card myClass"
  onClick={undefined}
  style={
    Object {
      "borderBottomColor": "#000000",
    }
  }
>
  <resilient-image
    props={
      Object {
        "alwaysShow": true,
        "className": "imageContainer",
        "showSpinner": true,
        "src": undefined,
      }
    }
  />
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
      }
    }
  >
    <div
      className="titleText"
      style={
        Object {
          "color": "#000000",
        }
      }
    >
      <span>
         
        myTitle
      </span>
    </div>
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given icon shown 1`] = `
<div
  className="card undefined"
  onClick={undefined}
  style={
    Object {
      "borderBottomColor": "#000000",
    }
  }
>
  <resilient-image
    props={
      Object {
        "alwaysShow": true,
        "className": "imageContainer",
        "showSpinner": true,
        "src": undefined,
      }
    }
  />
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
      }
    }
  >
    <div
      className="titleText"
      style={
        Object {
          "color": "#000000",
        }
      }
    >
      <span>
        <i
          className="mdi mdi-monitor"
        />
         
        myTitle
      </span>
    </div>
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given image 1`] = `
<div
  className="card undefined"
  onClick={undefined}
  style={
    Object {
      "borderBottomColor": "#000000",
    }
  }
>
  <resilient-image
    props={
      Object {
        "alwaysShow": true,
        "className": "imageContainer",
        "showSpinner": true,
        "src": "myImage",
      }
    }
  />
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
      }
    }
  >
    <div
      className="titleText"
      style={
        Object {
          "color": "#000000",
        }
      }
    >
      <span>
         
        myTitle
      </span>
    </div>
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given style 1`] = `
<div
  className="card undefined"
  onClick={undefined}
  style={
    Object {
      "borderBottomColor": "#000000",
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
>
  <resilient-image
    props={
      Object {
        "alwaysShow": true,
        "className": "imageContainer",
        "showSpinner": true,
        "src": undefined,
      }
    }
  />
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
      }
    }
  >
    <div
      className="titleText"
      style={
        Object {
          "color": "#000000",
        }
      }
    >
      <span>
         
        myTitle
      </span>
    </div>
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given title 1`] = `
<div
  className="card undefined"
  onClick={undefined}
  style={
    Object {
      "borderBottomColor": "#000000",
    }
  }
>
  <resilient-image
    props={
      Object {
        "alwaysShow": true,
        "className": "imageContainer",
        "showSpinner": true,
        "src": undefined,
      }
    }
  />
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(transparent, rgba(0, 0, 0, 0.9))",
      }
    }
  >
    <div
      className="titleText"
      style={
        Object {
          "color": "#000000",
        }
      }
    >
      <span>
         
        myTitle
      </span>
    </div>
  </div>
</div>
`;
